### ISSUE TITLE
CSS Inlining is slow when inlining wildcard styles on a large document

### ISSUE THREAD

**Issue by vlymar** [danschultzer/premailex#90]
Hi @danschultzer, I've been investigating a scenario in which premailex takes a very long time to finish inlining CSS and I had some questions.

The scenario is that the input html has a number of font styles in the `<head>` that have `*` selectors, and the email body is large (thousands of tags). In this scenario Premailex takes ~30s to inline css. I generated a flamegraph of the inlining process for a much smaller doc (inlining a wildcard style) and saw that ~75% of the time is spent traversing the document.

I read through some of the code and saw that in this case a lot of the traversal is probably coming from [this code](https://github.com/danschultzer/premailex/blob/dde02cd8143cb80b4f08e5cea303fa7d59cd68e2/lib/premailex/html_inline_styles.ex#L129-L137). 

I was wondering why use `Util.traverse_until_first` for every matching element rather than using something like [`Floki.find_and_update`](https://hexdocs.pm/floki/Floki.html#find_and_update/3) to match and update elements in one pass? The extra `traverse_until_first` ends up being very expensive when the selector is `*` as for every single element you have to re-traverse the document until you find it. In my quick test, switching to `Floki.find_and_update` made inlining ~7.5x faster for this particular scenario. I haven't dug super deep into the code so I could be totally wrong on the problem.

Is the issue that meeseeks doesn't support updating, so premailex needs to own the updating logic to remain compatible with both Floki and Meeseeks?

Thanks for your time!

**Comment by danschultzer** [comment]
Yeah, I see that. Did you find a solution for this? I would like to explore how this can be optimized, but realistically I likely won't have much time to explore this, so any code examples or PR would be appreciated. I think the reason it works the way it works now was that Meeseeks and Floki was pretty limited in what they could do 5-6 years ago.

**Comment by danschultzer** [comment]
This took a very long time, but I got it fixed by rolling out my own CSS parser and tree traversal. You should see a speedup of 100x or more, with giant 10k node HTML trees with only wildcard selectors I still see avg of around 145ms to inline styles.
